src/plugins/workspace/public/services/workspace_collaborator_types_service.ts

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { BehaviorSubject } from 'rxjs';
+import { i18n } from '@osd/i18n';
+
+// Use string literal avoid import enum from workspace self
+type WorkspaceCollaboratorPermissionSettingType = 'user' | 'group';
+
+export interface WorkspaceCollaboratorType {
+  id: string;
+  name: string;
+  pluralName: string;
+  // Will be assigned to this type in the final ACL permissions object
+  permissionSettingType: WorkspaceCollaboratorPermissionSettingType;
+  modal: {
+    title: string;
+    description?: string;
+    // Will use name with ID suffix for fallback
+    inputLabel?: string;
+    inputDescription?: string;
+    // Will use name with ID suffix for fallback
+    inputPlaceholder?: string;
+  };
+  instruction?: {
+    title: string;
+    detail: string;
+    link?: string;
+  };
+  // To match wether passed collaborator match this collaborator type
+  collaboratorMatcher?: (collaborator: {
+    type: WorkspaceCollaboratorPermissionSettingType;
+    userOrGroupId: string;
+  }) => boolean;
+}
```

Comment:
```suggestion
export interface WorkspaceCollaboratorType {
  id: string;
  name: string;
  onAdd: () => any
}

// `onAdd` function will be called when the dropdown menu been clicked
// It will show a modal, for example something like
{
  onAdd: () => coreStart.overlays.openModal(mount(<AddCollaboratorModal permissionType={} title="" {...} />))
}
```

This way, we maximize the flexibility of how the add collaborator modal should be look like, at the same time, it keeps the API simple. What do you think?

Reply:
I guess `collaboratorMatcher` is needed either way because it will be used to populate the type column in collaborator table.

If we are gonna to use onAdd, I guess we need to add some callback functions like

```
onAdd: (props: { onChange: (value: valueTypeForCollaborator) => void }) => ...
```

Reply:
Thanks @SuZhou-Joe what will `onChange` do?

Reply:
I think `valueTypeForCollaborator` should be an array. Will refactor code like this and include the AddCollaboratorModal in this PR.

Reply:
If we use a modal to maximize the flexibility, we need a callback to tell the parent component that user has added a user id. Or I do not know how we can communicate with parent component.